system/src/system_power_manager.cpp

```diff
@@ -148,6 +152,23 @@ void PowerManager::handleUpdate() {
     uint8_t vbus_stat = status >> 6;
     switch (vbus_stat) {
       case 0x01:
+#if HAL_INCREASE_CHARGING_CURRENT_WHEN_POWERED_BY_VUSB
```

Comment:
This logic doesn't cover the case when the device is powered by both VIN and through USB port and then during the runtime the USB cable is disconnected or reconnected.

We need to subscribe to changes of the USB state and manually trigger DPDM detection when the state changes to `HAL_USB_STATE_DETACHED` or when it changes from `HAL_USB_STATE_DETACHED` to any other. I propose we add a HAL function to register a callback to received USB state changes:

```c++
typedef void (*HAL_USB_State_Callback)(HAL_USB_State state, void* data);
int HAL_USB_Set_State_Change_Callback(HAL_USB_State_Callback cb, void* context, void* reserved);
```

Reply:
Now, it supports automatically increasing charging current when powered both by VUSB and USB host. Test several times to plug out and plug in the USB cable, the charging current changes correctly.